### PR TITLE
Elasticsearch exporter: event publishing

### PR DESCRIPTION
**Pull request by urso**
**Description:** Implement per document retry behavior and a set of tests verifying that the retry behavior is correctly configured and executed in conjunction with the BulkIndexer.

The change introduces helpers for creating a fake Elasticsearch endpoint that accepts bulk requests. This allows us to test the retry behavior for different failure scenarios.

**Link to tracking Issue:** #1800 

**Testing:** `TestExporter_PushEvent` tests the retry behavior for multiple error scenarios like http requests failing, or single documents being rejected due to overload.

**Documentation:** None. No user visible changes

### CODE REVIEW

**Comment by gramidt** [comment]
Codecov should pass once #2391 and #2389 are merged.

**Review by kbrockhoff — APPROVED** [review]

**Review by gramidt — APPROVED** [review]

**Comment by urso** [comment]
hm, CI failed in unrelated unit test:

```
ok  	github.com/open-telemetry/opentelemetry-collector-contrib/receiver/awsxrayreceiver/internal/translator	0.079s	coverage: 97.7% of statements
--- FAIL: TestSocketReadIrrecoverableNetError (0.01s)
    obsreporttest.go:160: 
        	Error Trace:	obsreporttest.go:160
        	            				obsreporttest.go:113
        	            				poller_test.go:301
        	Error:      	could not find tags
        	Test:       	TestSocketReadIrrecoverableNetError
        	Messages:   	wantTags: [{{receiver} TestSocketReadIrrecoverableNetError} {{transport} udp}] in rows []
FAIL
```

CircleCI does not allow me to rerun the test (without pushing a change)...

**Comment by gramidt** [comment]
@urso - Yeah, not being able to re-run the test from CircleCI is a pain point; however, you can `--amend --allow-empty` to your previous commit and force push to re-kick it off.

**Comment by tigrannajaryan** [comment]
@axw @simitt @jalvz you are codeowners of elasticexporter. Please review this PR.

@urso please add @axw @simitt @jalvz (or other Elastic contributos) as codeowners for elasticsearchexporter directory.

**Comment by urso** [comment]
> (or other Elastic contributos) as codeowners for elasticsearchexporter directory.

PR adding Codeowners for the elasticsearchexporter directory: #2462

**Comment by axw** [comment]
I will leave it to the exporter/elasticsearchexporter codeowners to review. @urso please let me know if you'd like additional review.

File: exporter/elasticsearchexporter/exporter.go
Comment (faec) [inline exporter/elasticsearchexporter/exporter.go:205]: Naming clarity: I generally think of "attempts" and "retries" as being off by 1 -- i.e. if the max "attempts" is 3, that means you only "retry" (try again) twice. This convention isn't universal, but it seems confusing to use both terms in close proximity. It would be nice to stick with the same term in both structs and local variables if possible, or to note the discrepancy at the definition sites if not.

File: exporter/elasticsearchexporter/exporter.go
Comment (faec) [inline exporter/elasticsearchexporter/exporter.go:102]: It's not obvious to me how `maxRetries` and `Retry.Enabled` interact. Is it possible to have `Retry.Enabled` set to false while still setting `maxRetries > 1`? If so, there should probably be a check for that here, or a check in `newElasticSearchClient` that overwrites the configured `maxRetries` if `Retry.Enabled` is false. If not, a comment explaining why this can't arise would be nice :-)

**Review by faec — COMMENTED** [review]
Looks good to me, some small comments about naming and the details of the retry logic

File: exporter/elasticsearchexporter/exporter.go
Comment (urso) [reply exporter/elasticsearchexporter/exporter.go:204]: Nice catch. Yes, the second should not be not.

File: exporter/elasticsearchexporter/exporter.go
Comment (urso) [reply exporter/elasticsearchexporter/exporter.go:205]: The structs is not defined in the exporter but the go-elasticsearch dependency. Debugging the tests I learned that the library starts the count with 1 (if 0 is passed the library will default to 3 retries instead of disabling retries). Will add a clarifying comment to `maxAttempts`. After testing I figured that the behavior is more similar to an attempt, which is what I named the variable with.

File: exporter/elasticsearchexporter/exporter.go
Comment (urso) [reply exporter/elasticsearchexporter/exporter.go:102]: if `Retry.Enabled` is false, then events should be dropped immediately on error. I used to have a check somewhere. Let me check and maybe add more tests if necessary to verify `Retry.Enabled=false` really disables all retry attempts.

Comment (urso) [reply exporter/elasticsearchexporter/exporter.go:102]: renaming `maxRetries` to `maxAttempts`. The `maxAttempts` is actually configured in `newExporter`. In case retry is disabled we set `maxAttempts = 1`. On Failure the predicate always evaluates to false thanks to the proper initialization of `maxAttempts`.

Note: the newElasticsearchClient configures the HTTP client. If the http request did succeed, we will have the bulk indexer report the per item status to `OnSuccess` and `OnFailure`. The client handles the HTTP retries for us, but the per item retries we have to handle ourselves.

**Pull request merged**
